### PR TITLE
Add configurable optimizer utilities for training loop

### PR DESCRIPTION
**Pull request by ChangLiu-doc**
NEW: Moved optimizer from training loop to optimisation and using the optimiser_factory.py for the function. 
          Fixed the problems listed by Maryam.

OLD: This PR adds configurable optimizer utilities for the shared training loop.

The optimizer system is now modularised similarly to the scheduler utilities. Instead of hard-coding optimizers directly inside the training configuration builder, optimizers are now created through a dedicated utility module.

The initial supported optimizers are:

Adam
AdamW
SGD
RMSprop
Adagrad

This allows future experimentation and hyperparameter search without modifying the core training loop logic.

This PR focuses only on optimizer modularisation and intentionally does not include scheduler changes, as scheduler utilities are handled separately.

The new optimizer utilities are designed to align with the existing scheduler utility architecture for consistency across the training infrastructure.

### CODE REVIEW

**Review by Mars-Martiny — CHANGES_REQUESTED** [review]
I requested that the file adding the configurable optimisers not be placed in the training_loop folder to avoid making it larger than it already is.
Please adjust accordingly.

File: leaderboard/leaderboard.csv
Comment (Mars-Martiny) [inline leaderboard/leaderboard.csv:1]: FIX - You are overwriting/deleting the previous leaderboard logs.

File: modules/training_loop/__init__.py
Comment (Mars-Martiny) [inline modules/training_loop/__init__.py:1]: FIX - Functions `create_optimizer` and `normalise_optimizer_config` do not need to be imported in the `__init__.py` file, they are not called anywhere within?

FIX - Comments should be adjusted to list expected (and acceptable) values for optimiser and optimiser_args.

File: modules/training_loop/optimizer.py
Comment (Mars-Martiny) [inline modules/training_loop/optimizer.py:1]: FIX - `ValueError` should be avoided, instead a catch should be implemented such it is not necessary to raise an error.

FIX - Type detection for optimizer objects is too permissive.

FIX - No safety net for invalid optimiser arguments passed (e.g. passing momentum value to Adam optimiser), nothing to inform user of source if this TypeError is raised.

**Review by Mars-Martiny — CHANGES_REQUESTED** [review]
This PR includes all the files related to your model implementations, this is too much to review in one PR and makes it easier for check everything.

I left comments listing fixes that should be made. I focused on the files regarding optimiser so I cannot say much about the model implementations at this point in time.

File: leaderboard/leaderboard.csv
Comment (ChangLiu-doc) [reply leaderboard/leaderboard.csv:1]: I have aborted the change made and kept the original content of the leaderboard.

File: modules/training_loop/__init__.py
Comment (ChangLiu-doc) [reply modules/training_loop/__init__.py:1]: Have deleted Functions optimizer and normalise_optimizer_config from __init__.py.

Has adjusted the comments accordingly.

File: modules/training_loop/optimizer.py
Comment (ChangLiu-doc) [reply modules/training_loop/optimizer.py:1]: optimizer has been moved outside the training loop and in the optimizer itself: 
1. ValueError has been replaced by return None. For example:
if optimiser_class is None:
    print(
        f"[Optimiser Factory] Unsupported optimiser '{name}'."
    )
    return None

2. changed to if isinstance(optimiser, torch.optim.Optimizer): 

3. TypeError has been replaced to
try:
    return optimiser_class(parameters, **args)

except TypeError as exc:
    print(
        f"[Optimiser Factory] Invalid arguments ..."
    )
    return None

**Comment by AndreHrs** [comment]
Hi Chang Liu, do you mind using this branch as reference https://github.com/Mars-Martiny/CITS5206_group1/tree/andre/rerun-leaderboard
Probably put the negative finding in the reports. This way you can justify the negative finding on what works and what did not work.
I'm not sure whether adding another model and bloating the code is necessary anymore
But if you still want to add it (just as reference for client). Please document it properly and report on it

Or we can just not merge the model, but at least the report should be merged (negative finding is equally important)

File: modules/training_loop/__init__.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 from .config import _build_train_config
 from .train_loop import train_model_loop
 
-__all__ = ["training", "_build_train_config", "train_model_loop"]
+__all__ = ["training", "_build_train_config", "train_model_loop", ]
```
Comment (Mars-Martiny) [inline modules/training_loop/__init__.py]:
```suggestion
__all__ = ["training", "_build_train_config", "train_model_loop"]
```

File: modules/training_loop/__init__.py
Comment (Mars-Martiny) [inline modules/training_loop/__init__.py]: This line should be adjusted accordingly to make it clear that it no longer accepts just an optimiser instance, rather it constructs one and the acceptable inputs.

**Review by Mars-Martiny — CHANGES_REQUESTED** [review]
Please remove the content relating to the bilstm and the svm, unless they were able to achieve higher than 50% accuracy it is not worth giving to the client. 
The tf-idf model is enough for comparison against as a simple model.

These bilstm and svm models should be documented (code documentation plus the max accuracy it was able to achieve) and they can be left in a branch for the client if they do want to explore it but it should not be placed in Dev (to keep it out of Main).
